bigtable/metrics.go

```diff
@@ -0,0 +1,479 @@
+			metricType: otelmetric.InstrumentKindHistogram,
+			unit:       "ns",
+			additionalAttributes: []string{
+				metricLabelKeyOperationStatus,
+				metricLabelKeyStreamingOperation,
+			},
+		},
+		metricNameRetryCount: {
+			desc:       "The number of additional RPCs sent after the initial attempt.",
+			metricType: otelmetric.InstrumentKindCounter,
+			additionalAttributes: []string{
+				metricLabelKeyOperationStatus,
+			},
+		},
+	}
+
+	noOpRecordFn = func() {}
+)
+
+type metricInfo struct {
+	desc                 string
+	metricType           otelmetric.InstrumentKind
+	unit                 string
+	additionalAttributes []string
+}
+
+type builtInMetricInstruments struct {
+	distributions map[string]metric.Float64Histogram // Key is metric name e.g. operation_latencies
+	counters      map[string]metric.Int64Counter     // Key is metric name e.g. retry_count
+}
+
+// Generates unique client ID in the format go-<random UUID>@<>hostname
+func generateClientUID() (string, error) {
+	hostname := "localhost"
+	hostname, err := os.Hostname()
+	if err != nil {
+		return "", err
+	}
+	return "go-" + uuid.NewString() + "@" + hostname, nil
+}
+
+type metricsConfigInternal struct {
+	exporter otelmetric.Exporter
+
+	project string
+
+	// Flag to indicate whether built-in metrics should be recorded and exported to GCM
+	builtInEnabled bool
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	// Contains one entry per meter provider
+	instruments []builtInMetricInstruments
+}
+
+func newMetricsConfigInternal(ctx context.Context, project, instance string, userProvidedConfig *metricsConfig, opts ...option.ClientOption) (*metricsConfigInternal, error) {
+	clientUID, err := generateClientUID()
+	if err != nil {
+		log.Printf("built-in metrics: generateClientUID failed: %v. Using empty string in the %v metric atteribute", err, metricLabelKeyClientUID)
+	}
+
+	// Create default meter provider
+	internalMetricsConfig := &metricsConfigInternal{
+		project: project,
+		clientAttributes: []attribute.KeyValue{
+			attribute.String(monitoredResLabelKeyProject, project),
+			attribute.String(monitoredResLabelKeyInstance, instance),
+			attribute.String(metricLabelKeyClientUID, clientUID),
+			attribute.String(metricLabelKeyClientName, clientName),
+		},
+		instruments:    []builtInMetricInstruments{},
+		builtInEnabled: builtInEnabledDefault,
+	}
+
+	defaultExporter, err := newMonitoringExporter(ctx, internalMetricsConfig.project, opts...)
+	if err != nil {
+		return nil, err
+	}
+	internalMetricsConfig.exporter = defaultExporter
+
+	// Create default meter provider
+	defaultMp := otelmetric.NewMeterProvider(
+		otelmetric.WithReader(
+			otelmetric.NewPeriodicReader(
+				defaultExporter,
+				otelmetric.WithInterval(defaultSamplePeriod),
+			),
+		),
+	)
+
+	userProvidedMeterProviders := []*otelmetric.MeterProvider{}
```

Comment:
I dont quite understand the usecase for a user giving us multiple MeterProviders.

Separately, I dont think you want to have separate MeterProviders for user and builtin metrics....I suspect it will create significant overhead because all of the aggregations will end up happening twice

Reply:
Updated the code to have single meterprovider